tools/PI/DevHome.PI/Views/BarWindowHorizontal.xaml.cs

```diff
@@ -606,15 +612,89 @@ private void WindowEx_WindowStateChanged(object sender, WindowState e)
+            // If, as part of being restored, we were in an expanded state, then make our window bigger (don't go back to collapsed mode)
+            if (_viewModel.ShowingExpandedContent && Height == FloatingHorizontalBarHeight)
+            {
+                Height = Settings.Default.ExpandedWindowHeight;
+            }
+        }
+        else if (e.Equals(WindowState.Maximized))
+        {
+            // If we're being maximized, expand our content
+            _viewModel.ShowingExpandedContent = true;
+        }
+    }
+
+    private Windows.Win32.UI.Shell.SUBCLASSPROC? _wndProc;
+
+    private void HookWndProc()
+    {
+        _wndProc = new Windows.Win32.UI.Shell.SUBCLASSPROC(NewWindowProc);
+        PInvoke.SetWindowSubclass(ThisHwnd, _wndProc, 456, 0);
+    }
+
+    private bool _isSnapped;
+    private bool _transitionFromSnapped;
+
+    private LRESULT NewWindowProc(HWND hWnd, uint msg, WPARAM wParam, LPARAM lParam, nuint uldSubclass, nuint dwRefData)
+    {
+        switch (msg)
+        {
+            case PInvoke.WM_WINDOWPOSCHANGING:
             {
-                // If we were in collapsed mode, then maximized, then expanded, then restored, be sure to set the height back to the expanded height
-                // (by default, Windows will set us to the collapsed height)
-                if (_viewModel.ShowingExpandedContent && Height == FloatingHorizontalBarHeight)
+                Windows.Win32.UI.WindowsAndMessaging.WINDOWPOS wndPos = Marshal.PtrToStructure<Windows.Win32.UI.WindowsAndMessaging.WINDOWPOS>(lParam);
+
+                // We only care about this message if it's triggering a resize
+                if (wndPos.flags.HasFlag(Windows.Win32.UI.WindowsAndMessaging.SET_WINDOW_POS_FLAGS.SWP_NOSIZE))
+                {
+                    break;
+                }
+
+                int floatingBarHeight = CommonHelper.MulDiv(FloatingHorizontalBarHeight, (int)this.GetDpiForWindow(), 96);
+
+                if (PInvoke.IsWindowArranged(hWnd))
+                {
+                    _isSnapped = true;
+                }
+                else
+                {
+                    if (_isSnapped)
+                    {
+                        _transitionFromSnapped = true;
+                        _isSnapped = false;
+                    }
+                }
+
+                if (!_isSnapped && !_viewModel.ShowingExpandedContent && wndPos.cy > floatingBarHeight)
                 {
-                    Height = Settings.Default.ExpandedWindowHeight;
+                    // Enforce our height limit if we're not showing expanded content and we're not being snapped
+                    wndPos.cy = CommonHelper.MulDiv(FloatingHorizontalBarHeight, (int)this.GetDpiForWindow(), 96);
+                    Marshal.StructureToPtr(wndPos, lParam, true);
+                    Debug.WriteLine("WM_WINDOWPOSCHANGING: Enforcing height limit " + _isSnapped + " " + _transitionFromSnapped);
                 }

```

Comment:
delete or swap to PI logger?

Reply:
Hmm... it doesn't look like we've made this easy. The logs are mildly interesting, especially if I've missed a case here. Moving this to the PI logger means it no longer shows up in DebugOutput, and so I need to look at yet another place to get this sort of tracing data. I've moved it to the PI logger for now, but something tells me I'd switch it back to Debug.WriteLine if I actually needed to debug scenarios. :()